Wire up the JS engine's memory reporting.

nnethercote opened the pull request:
SpiderMonkey provides an extremely fine-grained breakdown of memory
usage, but for Servo we aggregate the measurements into a small number
of coarse buckets, which seems appropriate for the current level of
detail provided by Servo's memory profiler. Sample output:

```
|      17.41 MiB -- url(file:///home/njn/moz/servo/../servo-static-suite/wikipedia/Guardians%20of%20the%20Galaxy%20(film)%20-%20Wikipedia,%20the%20free%20encyclopedia.html)
|          7.32 MiB -- js
|             3.07 MiB -- malloc-heap
|             3.00 MiB -- gc-heap
|                2.48 MiB -- used
|                0.34 MiB -- decommitted
|                0.09 MiB -- unused
|                0.09 MiB -- admin
|             1.25 MiB -- non-heap
```

Most of the changes are plumbing to get the script task communicating
with the memory profiler task.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/6572)

<!-- Reviewable:end -->

nnethercote committed:
Wire up the JS engine's memory reporting.

SpiderMonkey provides an extremely fine-grained breakdown of memory
usage, but for Servo we aggregate the measurements into a small number
of coarse buckets, which seems appropriate for the current level of
detail provided by Servo's memory profiler. Sample output:
```
|   10.99 MiB -- pages
|       7.75 MiB -- url(http://html5demos.com/worker)
|          4.63 MiB -- js
|             2.00 MiB -- gc-heap
|                0.94 MiB -- decommitted
|                0.92 MiB -- used
|                0.09 MiB -- unused
|                0.05 MiB -- admin
|             1.44 MiB -- malloc-heap
|             1.19 MiB -- non-heap
|          [...]
|       3.24 MiB -- url(http://html5demos.com/js/worker-cruncher.js)
|          3.24 MiB -- js
|             1.17 MiB -- malloc-heap
|             1.06 MiB -- non-heap
|             1.00 MiB -- gc-heap
|                0.69 MiB -- used
|                0.19 MiB -- decommitted
|                0.09 MiB -- unused
|                0.03 MiB -- admin
```
Most of the changes are plumbing to get the script and worker tasks
communicating with the memory profiler task.

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!

nnethercote commented:
@jdm, this review has your name all over it :)

I'm actually going to have to land patches in four repositories before this is done.
- An API change to SpiderMonkey;
- Backport that to mozjs;
- Add appropriate glue to rust-mozjs;
- Use the glue in Servo (i.e. this change).

I'm putting this change up for review first because it's driving the other changes. If any modifications are needed here it'll be best to work them out here first and then back-propagate them as necessary.

I'm a little worried about testing all this. I can do a certain amount of local testing (e.g. I've done Linux and Mac) but I'm worried about the following case.
- I get r+ on all the changes;
- I land the changes in the first three repos;
- I try to land the changes in the fourth repo (Servo) but I get some kind of CI test failure that doesn't occur locally;
- I potentially then have to diagnose and fix that failure by modifying my changes in all four repos.

That would be truly painful, but I can't see a way to CI-test all of my changes in advance. Hopefully it won't come to that.

michaelwu commented:
I'd like to see the rust-mozjs changes.

nnethercote commented:
> I'm actually going to have to land patches in four repositories before this is done.
> 
> * An API change to SpiderMonkey;

This is https://bugzilla.mozilla.org/show_bug.cgi?id=1181452.

> * Backport that to mozjs;

This is https://github.com/servo/mozjs/pull/49

> * Add appropriate glue to rust-mozjs;

This is https://github.com/servo/rust-mozjs/pull/177

> * Use the glue in Servo (i.e. this change).

jdm commented:
It would be useful to hook up the worker threads to the memory reporting, too: http://mxr.mozilla.org/servo/source/components/script/dom/dedicatedworkerglobalscope.rs and http://mxr.mozilla.org/servo/source/components/script/dom/workerglobalscope.rs are where those are implemented.

-S-awaiting-review +S-needs-code-changes 

---

Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/script_task.rs, line 286 \[r1\]](https://reviewable.io:443/reviews/servo/servo/6572#-JthhZPdQsEdbBu3-3en)** ([raw file](https://github.com/servo/servo/blob/fea730b6ea8e5e8ca01b58bb0e62e8a434972fc6/components/script/script_task.rs#L286)):</sup>
Use no_jsmanaged_fields in trace.rs instead.

---

<sup>**[components/script/script_task.rs, line 295 \[r1\]](https://reviewable.io:443/reviews/servo/servo/6572#-JthhoONkmZ0THc_SGaa)** ([raw file](https://github.com/servo/servo/blob/fea730b6ea8e5e8ca01b58bb0e62e8a434972fc6/components/script/script_task.rs#L295)):</sup>
This is a bit misleading. A script task can contain multiple pages (eg. via same-origin iframes), and in the future we'll have a model of a script task containing all in-memory pages for a particular eTLD.

---

<sup>**[components/script/script_task.rs, line 528 \[r1\]](https://reviewable.io:443/reviews/servo/servo/6572#-JthkJgd4uJ1t6Nw0Xde)** ([raw file](https://github.com/servo/servo/blob/fea730b6ea8e5e8ca01b58bb0e62e8a434972fc6/components/script/script_task.rs#L528)):</sup>
If you move this into ScriptTask::start we don't need to store the reporter name as a member.

---

<sup>**[components/script/script_task.rs, line 731 \[r1\]](https://reviewable.io:443/reviews/servo/servo/6572#-Jthk8ycTBFt8oVGF-nY)** ([raw file](https://github.com/servo/servo/blob/fea730b6ea8e5e8ca01b58bb0e62e8a434972fc6/components/script/script_task.rs#L731)):</sup>
I'd prefer this in ScriptTask::start.

---

<sup>**[components/script_traits/lib.rs, line 149 \[r1\]](https://reviewable.io:443/reviews/servo/servo/6572#-Jthh32-jGq0FxHIHwUW)** ([raw file](https://github.com/servo/servo/blob/fea730b6ea8e5e8ca01b58bb0e62e8a434972fc6/components/script_traits/lib.rs#L149)):</sup>
We could extract this from the LoadData that's already passed instead.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6572)
<!-- Sent from Reviewable.io -->

nnethercote commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

<sup>**[components/script/script_task.rs, line 295 \[r1\]](https://reviewable.io:443/reviews/servo/servo/6572#-JthhoONkmZ0THc_SGaa)** ([raw file](https://github.com/servo/servo/blob/fea730b6ea8e5e8ca01b58bb0e62e8a434972fc6/components/script/script_task.rs#L295)):</sup>
Is it just the comment you dislike, or should I find a different way to distinguish/label a script task? Other memory (e.g. layout) is currently reported on a page basis, and it would be a shame if JS memory had to be done differently.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6572)
<!-- Sent from Reviewable.io -->

nnethercote commented:
> If you move this into ScriptTask::start we don't need to store the reporter name as a member.

Hmm, the same simplification could be applied to the existing reporters for the Layout and Paint tasks. I'll add that to my todo list.

nnethercote commented:
I've pushed my in-progress changes. I addressed all the comments except the one about hooking up workers; I'll do that tomorrow. And @jdm requested compartment-level reporting which I will consider (I'll need to think about compartments vs. zones...)

nnethercote commented:
I've hooked up workers now. I still haven't done anything about compartment-level reporting; my enthusiasm level for doing that is low.

jdm commented:
-S-awaiting-review +S-needs-code-changes 
I'd rather see this code merge as-is than have it wait for the compartment-level stuff if you're that unenthused.

---

Reviewed 4 of 5 files at r2, 6 of 6 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/bindings/global.rs, line 90 \[r3\]](https://reviewable.io:443/reviews/servo/servo/6572#-Ju6R5ck_Kh56LwnIAUa)** ([raw file](https://github.com/servo/servo/blob/1376326b5a360d60667ed4b1386e9b6637169529/components/script/dom/bindings/global.rs#L90)):</sup>
I'm pretty sure this will trigger if we spawn a worker inside of another worker.

---

<sup>**[components/script/script_task.rs, line 1030 \[r3\]](https://reviewable.io:443/reviews/servo/servo/6572#-Ju6RzcPktNWNYOoAlcb)** ([raw file](https://github.com/servo/servo/blob/1376326b5a360d60667ed4b1386e9b6637169529/components/script/script_task.rs#L1030)):</sup>
I think `for it_page in &self.root_page()` should work.

---

<sup>**[components/script/script_task.rs, line 1031 \[r3\]](https://reviewable.io:443/reviews/servo/servo/6572#-Ju6S63vXNc8Ev8W9t0Q)** ([raw file](https://github.com/servo/servo/blob/1376326b5a360d60667ed4b1386e9b6637169529/components/script/script_task.rs#L1031)):</sup>
`urls.push(it_page.document().url().serialize());`

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6572)
<!-- Sent from Reviewable.io -->

nnethercote commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

<sup>**[components/script/script_task.rs, line 1030 \[r3\]](https://reviewable.io:443/reviews/servo/servo/6572#-Ju6RzcPktNWNYOoAlcb)** ([raw file](https://github.com/servo/servo/blob/1376326b5a360d60667ed4b1386e9b6637169529/components/script/script_task.rs#L1030)):</sup>
I tried that previously, but rustc says: "error: the trait `core::iter::Iterator` is not implemented for the type `&alloc::rc::Rc<page::Page>`". I also tried without the `&` and with a '*' and got much the same error. `Page` does not implement `Iterator`, but `Rc<Page>` does implement `IterablePage`, which requires the use of `iter()`.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6572)
<!-- Sent from Reviewable.io -->

nnethercote commented:
Ok! All review comments are addressed. But this still needs to wait for the mozjs and rust-mozjs changes to land first.

jdm commented:
-S-awaiting-review +S-blocked-on-external 

---

Reviewed 4 of 4 files at r4.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6572)
<!-- Sent from Reviewable.io -->

bors-servo commented:
:umbrella: The latest upstream changes (presumably #6626) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
-S-blocked-on-external -S-awaiting-review +S-needs-code-changes 

---

Reviewed 8 of 8 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

<sup>**[ports/cef/Cargo.lock, line 629 \[r5\]](https://reviewable.io:443/reviews/servo/servo/6572#-JuDnD0hfnbyKr18lKKc)** ([raw file](https://github.com/servo/servo/blob/d992f98b0102477ca406542d9130e77487bb64b4/ports/cef/Cargo.lock#L629)):</sup>
You'll want to run `./mach update-cargo -p rust-mozjs` and `./mach update-cargo -p mozjs`.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6572)
<!-- Sent from Reviewable.io -->

nnethercote commented:
> You'll want to run ./mach update-cargo -p rust-mozjs and ./mach update-cargo -p mozjs.

I had a hell of a time with this. Eventually I worked out it should be "./mach update-cargo -p mozjs_sys" and "./mach update-cargo -p js".

metajack commented:
I could have sworn it used to work with dashes.

metajack commented:
@alexcrichton did the cargo behavior change?

alexcrichton commented:
Nah this hasn't changed recently to the best of my knowledge.

nnethercote commented:
I don't know what dashes have to do with it...

AIUI the problem is that the name used within Servo for the crates(?) don't match the repo names, and "./mach cargo-update -p" needs the former rather than the latter.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #6591) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit d70fdc5 has been approved by `jdm`

<!-- @bors-servo r=jdm d70fdc56c9fd4e3d82c11638b463086c1bc614b8 -->

bors-servo commented:
:hourglass: Testing commit d70fdc5 with merge 800c090...

bors-servo commented:
:broken_heart: Test failed - [mac1](http://build.servo.org/builders/mac1/builds/2522)

jdm commented:
```
17:35.11 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:27325) Full command: /Users/servo/buildbot/slave/mac1/build/target/debug/servo --cpu --hard-fail -u Servo/wptrunner -z http://localhost:8000/workers/semantics/multiple-workers/002.html
(pid:27325) "thread 'Memory profiler' panicked at 'RegisterReporter: 'worker-reporter-0' name is already in use', /Users/servo/buildbot/slave/mac1/build/components/profile/mem.rs:87"
17:35.11 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:27325) "stack backtrace:"
17:35.11 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:27325) "   1:        0x10f99f9d5 - ALERT: RESULT: {"tests":[{"name":"creating 3 sibling dedicated workers","status":0,"message":null,"stack":null}],"status":0,"message":null,"stack":null}"
17:35.11 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:27325) "sys::backtrace::write::hde88a3e3a343e9f8zms"
17:35.11 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:27325) "   2:        0x10f9a2fc5 - panicking::on_panic::h8c73c8f1e33f6d60hIw"
17:35.11 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:27325) "   3:        0x10f990ab2 - rt::unwind::begin_unwind_inner::h781ff1cd39e7f47etqw"
17:35.11 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:27325) "   4:        0x10d2e1385 - rt::unwind::begin_unwind::h389154536123609620"
17:35.11 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:27325) "   5:        0x10d2d7853 - mem::Profiler::handle_msg::hef9799da9a1e7247Zca"
17:35.11 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:27325) "   6:        0x10d2ced4e - mem::Profiler::start::h457c0bdc01213071Aca"
17:35.12 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:27325) "   7:        0x10d2ceb45 - mem::Profiler::create::closure.5520"
17:35.12 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:27325) "   8:        0x10d2cea1e - task::spawn_named::closure.5514"
17:35.12 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:27325) "   9:        0x10d2ce975 - boxed::F.FnBox<A>::call_box::h3233880008695368661"
17:35.12 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:27325) "  10:        0x10d2cb3b0 - boxed::Box<FnBox<A, Output $u3d$$u20$R$GT$$u2b$$u20$Send$u20$$u2b$$u20$$u27$a$GT$.FnOnce$LT$A$GT$::call_once::h12673241061259117211"
17:35.12 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:27325) "  11:        0x10d2cae22 - thread::Builder::spawn_inner::closure.5367"
17:35.12 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:27325) "  12:        0x10d2cad9e - rt::unwind::try::try_fn::__rust_abi::h12586371467759283055"
17:35.12 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:27325) "  13:        0x10d2cad39 - rt::unwind::try::try_fn::h12586371467759283055"
17:35.12 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:27325) "  14:        0x10f9a49c8 - rust_try_inner"
17:35.13 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:27325) "  15:        0x10f9a49b5 - rust_try"
17:35.13 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:27325) "  16:        0x10f99df75 - rt::unwind::try::inner_try::h9e58480e93edc0cdmmw"
17:35.13 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:27325) "  17:        0x10d2cac98 - rt::unwind::try::h17192558137596237202"
17:35.13 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:27325) "  18:        0x10d2caad1 - thread::Builder::spawn_inner::closure.5317"
17:35.13 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:27325) "  19:        0x10d2cb73d - boxed::F.FnBox<A>::call_box::h14907357354884052942"
17:35.14 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:27325) "  20:        0x10f9a198d - sys::thread::Thread::new::thread_start::hf153915745363fa7wKv"
17:35.14 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:27325) "  21:     0x7fff94cef267 - _pthread_body"
17:35.14 PROCESS_OUTPUT: Thread-TestrunnerManager-2 (pid:27325) "  22:     0x7fff94cef1e4 - _pthread_start"
```
Whoops, we should probably fix that. Let's use the worker id (http://mxr.mozilla.org/servo/source/components/script/dom/worker.rs#86) and append that to the pipeline id so we get globally-unique names.

jdm commented:
Actually, that's probably not good enough, since we've got separate "next_worker_id" values in Window and WorkerGlobalScope, so a name like "worker-reporter-{pipeline}-{worker_id}" yields "worker-reporter-0-0" for a toplevel worker, then "worker-reporter-0-0" for the first subworker too.

jdm commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit 7429b90 has been approved by `jdm`

<!-- @bors-servo r=jdm 7429b90e02effd086681196050f400b2d1836903 -->

bors-servo commented:
:hourglass: Testing commit 7429b90 with merge 2b0bdbe...

bors-servo commented:
:broken_heart: Test failed - [mac1](http://build.servo.org/builders/mac1/builds/2531)

nnethercote commented:
websockets/constructor/004.html failed with "thread 'LayoutTask PipelineId(0)' panicked at 'could not initialize thread_rng: Too many open files (os error 24)', src/libstd/rand/mod.rs:161"

Although my latest revision did add calls to rand::random() this doesn't feel like it's fundamentally my fault :(

nnethercote commented:
Looks like taht thread_rng panic has been seen before in #6350.

mbrubeck commented:
@bors-servo retry

bors-servo commented:
:zap: Previous build results for [android](http://build.servo.org/builders/android/builds/2305), [gonk](http://build.servo.org/builders/gonk/builds/1458), [linux1](http://build.servo.org/builders/linux1/builds/2305), [linux2](http://build.servo.org/builders/linux2/builds/2308), [linux3](http://build.servo.org/builders/linux3/builds/461), [mac2](http://build.servo.org/builders/mac2/builds/2536), [mac3](http://build.servo.org/builders/mac3/builds/489) are reusable. Rebuilding only [mac1](http://build.servo.org/builders/mac1/builds/2531)...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/2305), [gonk](http://build.servo.org/builders/gonk/builds/1458), [linux1](http://build.servo.org/builders/linux1/builds/2305), [linux2](http://build.servo.org/builders/linux2/builds/2308), [linux3](http://build.servo.org/builders/linux3/builds/461), [mac1](http://build.servo.org/builders/mac1/builds/2532), [mac2](http://build.servo.org/builders/mac2/builds/2536), [mac3](http://build.servo.org/builders/mac3/builds/489)
